[BUG] test_assemble_meta fails with Python 3.14.0b1

- [x] I've checked [docs](https://rich.readthedocs.io/en/latest/introduction.html) and [closed issues](https://github.com/Textualize/rich/issues?q=is%3Aissue+is%3Aclosed) for possible solutions.
- [x] I can't find my issue in the [FAQ](https://github.com/Textualize/rich/blob/master/FAQ.md).

**Describe the bug**

With Python 3.14 and various fixes from open PRs applied, I'm down to one failing test with beta 1.
I try to build rich for Fedora Linux.

```
______________________________ test_assemble_meta ______________________________

    def test_assemble_meta():
        text = Text.assemble("foo", ("bar", "bold"), meta={"foo": "bar"})
        assert str(text) == "foobar"
>       assert text._spans == [Span(3, 6, "bold"), Span(0, 6, Style(meta={"foo": "bar"}))]
E       AssertionError: assert [Span(3, 6, 'bold'), Span(0, 6, Style(meta={'foo': 'bar'}))] == [Span(3, 6, 'bold'), Span(0, 6, Style(meta={'foo': 'bar'}))]
E         
E         At index 1 diff: Span(0, 6, Style(meta={'foo': 'bar'})) != Span(0, 6, Style(meta={'foo': 'bar'}))
E         
E         Full diff:
E           [
E               Span(3, 6, 'bold'),
E               Span(0, 6, Style(meta={'foo': 'bar'})),
E           ]

tests/test_text.py:846: AssertionError
=========================== short test summary info ============================
FAILED tests/test_text.py::test_assemble_meta - AssertionError: assert [Span(3, 6, 'bold'), Span(0, 6, Style(meta={'foo': 'bar'}))] == [Span(3, 6, 'bold'), Span(0, 6, Style(meta={'foo': 'bar'}))]
  
  At index 1 diff: Span(0, 6, Style(meta={'foo': 'bar'})) != Span(0, 6, Style(meta={'foo': 'bar'}))
  
  Full diff:
    [
        Span(3, 6, 'bold'),
        Span(0, 6, Style(meta={'foo': 'bar'})),
    ]
```
**Platform**
<details>
<summary>Click to expand</summary>

It's the test run during the build of Fedora Linux.

</details>

--- a/.github/workflows/pythonpackage.yml
+++ b/.github/workflows/pythonpackage.yml
@@ -9,7 +9,7 @@ jobs:
       fail-fast: false
       matrix:
         os: [windows-latest, ubuntu-latest, macos-latest]
-        python-version: ["3.8", "3.9", "3.10", "3.11", "3.12", "3.13", "3.14"]
+        python-version: ["3.9", "3.10", "3.11", "3.12", "3.13", "3.14"]
         exclude:
           - { os: windows-latest, python-version: "3.13" }
     defaults:

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,17 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+
+# Unreleased
+
+### Changed
+
+- Dropped support for Python3.9
+
+### Fixed
+
+- Fixed empty print ignoring the `end` parameter
+
 ## [14.3.4] - 2026-04-11
 
 ### Changed

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -15,7 +15,6 @@ classifiers = [
     "Operating System :: Microsoft :: Windows",
     "Operating System :: MacOS",
     "Operating System :: POSIX :: Linux",
-    "Programming Language :: Python :: 3.8",
     "Programming Language :: Python :: 3.9",
     "Programming Language :: Python :: 3.10",
     "Programming Language :: Python :: 3.11",
@@ -28,7 +27,7 @@ include = ["rich/py.typed"]
 
 
 [tool.poetry.dependencies]
-python = ">=3.8.0"
+python = ">=3.9.0"
 pygments = "^2.13.0"
 ipywidgets = { version = ">=7.5.1,<9", optional = true }
 markdown-it-py = ">=2.2.0"

--- a/rich/console.py
+++ b/rich/console.py
@@ -2223,7 +2223,7 @@ class Console:
 
     def save_text(
         self,
-        path: Union[str, PathLike],
+        path: Union[str, PathLike[str]],
         *,
         clear: bool = True,
         styles: bool = False,
@@ -2320,7 +2320,7 @@ class Console:
 
     def save_html(
         self,
-        path: Union[str, PathLike],
+        path: Union[str, PathLike[str]],
         *,
         theme: Optional[TerminalTheme] = None,
         clear: bool = True,
@@ -2605,7 +2605,7 @@ class Console:
 
     def save_svg(
         self,
-        path: Union[str, PathLike],
+        path: Union[str, PathLike[str]],
         *,
         title: str = "Rich",
         theme: Optional[TerminalTheme] = None,

--- a/rich/console.py
+++ b/rich/console.py
@@ -2223,7 +2223,7 @@ class Console:
 
     def save_text(
         self,
-        path: Union[str, PathLike[str]],
+        path: Union[str, PathLike],
         *,
         clear: bool = True,
         styles: bool = False,
@@ -2320,7 +2320,7 @@ class Console:
 
     def save_html(
         self,
-        path: Union[str, PathLike[str]],
+        path: Union[str, PathLike],
         *,
         theme: Optional[TerminalTheme] = None,
         clear: bool = True,
@@ -2605,7 +2605,7 @@ class Console:
 
     def save_svg(
         self,
-        path: Union[str, PathLike[str]],
+        path: Union[str, PathLike],
         *,
         title: str = "Rich",
         theme: Optional[TerminalTheme] = None,

--- a/rich/console.py
+++ b/rich/console.py
@@ -7,6 +7,7 @@ from datetime import datetime
 from functools import wraps
 from itertools import islice
 from math import ceil
+from os import PathLike
 from time import monotonic
 from types import FrameType, ModuleType, TracebackType
 from typing import (
@@ -29,7 +30,6 @@ from typing import (
     cast,
     runtime_checkable,
 )
-from os import PathLike
 
 from rich._null_file import NULL_FILE
 
@@ -1687,7 +1687,10 @@ class Console:
             new_line_start (bool, False): Insert a new line at the start if the output contains more than one line. Defaults to ``False``.
         """
         if not objects:
-            objects = (NewLine(),)
+            if end == "\n":
+                objects = (NewLine(),)
+            else:
+                objects = ("",)
 
         if soft_wrap is None:
             soft_wrap = self.soft_wrap
@@ -2218,7 +2221,13 @@ class Console:
                 del self._record_buffer[:]
         return text
 
-    def save_text(self, path: Union[str, PathLike[str]], *, clear: bool = True, styles: bool = False) -> None:
+    def save_text(
+        self,
+        path: Union[str, PathLike[str]],
+        *,
+        clear: bool = True,
+        styles: bool = False,
+    ) -> None:
         """Generate text from console and save to a given location (requires record=True argument in constructor).
 
         Args:

--- a/tests/test_console.py
+++ b/tests/test_console.py
@@ -192,6 +192,12 @@ def test_print() -> None:
     assert console.file.getvalue() == "foo\n"
 
 
+def test_print_empty_with_end() -> None:
+    console = Console(file=io.StringIO(), color_system="truecolor")
+    console.print(end="!")
+    assert console.file.getvalue() == "!"
+
+
 def test_print_multiple() -> None:
     console = Console(file=io.StringIO(), color_system="truecolor")
     console.print("foo", "bar")

